crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -1,12 +1,119 @@
-use crate::{verbatim_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
-use rustpython_parser::ast::ExprBinOp;
+use crate::comments::trailing_comments;
+use crate::prelude::*;
+use crate::FormatNodeRule;
+use ruff_formatter::{write, FormatOwnedWithRule, FormatRefWithRule};
+use ruff_python_ast::node::{AnyNodeRef, AstNode};
+use rustpython_parser::ast::{
+    Constant, Expr, ExprAttribute, ExprBinOp, ExprConstant, ExprUnaryOp, Operator, Unaryop,
+};
 
 #[derive(Default)]
 pub struct FormatExprBinOp;
 
 impl FormatNodeRule<ExprBinOp> for FormatExprBinOp {
     fn fmt_fields(&self, item: &ExprBinOp, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [verbatim_text(item.range)])
+        let ExprBinOp {
+            left,
+            right,
+            op,
+            range: _,
+        } = item;
+
+        let needs_space = !is_simple_power_expression(item);
+
+        let before_operator_space = if needs_space {
+            soft_line_break_or_space()
+        } else {
+            soft_line_break()
+        };
+
+        let comments = f.context().comments().clone();
+        let operator_comments = comments.dangling_comments(item.as_any_node_ref());
+
+        write!(
+            f,
+            [
+                left.format(),
```

Comment:
This implementation heavily relies on recursion. We may want to explore to manually unroll left-balanced binary expressions because this could stack-overflow for very deep binary expressions (but so could any of our analyzer infrastructure).